components/tidb_query/src/codec/mysql/time/mod.rs

```diff
@@ -57,605 +48,1156 @@ const MONTH_NAMES_ABBR: &[&str] = &[
     "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
 ];
 
-#[inline]
-fn zero_time(tz: &Tz) -> DateTime<Tz> {
-    tz.timestamp(ZERO_TIMESTAMP, 0)
+fn is_leap_year(year: u32) -> bool {
+    year & 3 == 0 && (year % 100 != 0 || year % 400 == 0)
```

Comment:
Postgresql uses:
```
#define isleap(y) (((y) % 4) == 0 && (((y) % 100) != 0 || ((y) % 400) == 0))
```

So I don't think `year & 3` is better than `year % 4`. The compiler will help you to optimize complementation.

Reply:
Here is a stackoverflow link:
https://stackoverflow.com/questions/22446425/do-c-c-compilers-such-as-gcc-generally-optimize-modulo-by-a-constant-power-of

Reply:
I don't think so. This code compile by `clang version 8.0.0-3` with `-O3`
```c
int main() {
    int a;
    scanf("%d", &a);
    int b = a % 4;
    printf("%d", b);
    
    int c = a&3;
    printf("%d", c);
}
```
```asm
  40114d:	e8 ee fe ff ff       	callq  401040 <__isoc99_scanf@plt>
  401152:	8b 74 24 04          	mov    0x4(%rsp),%esi
  401156:	89 f0                	mov    %esi,%eax
  401158:	c1 f8 1f             	sar    $0x1f,%eax
  40115b:	c1 e8 1e             	shr    $0x1e,%eax
  40115e:	01 f0                	add    %esi,%eax
  401160:	83 e0 fc             	and    $0xfffffffc,%eax
  401163:	29 c6                	sub    %eax,%esi
  401165:	bf 04 20 40 00       	mov    $0x402004,%edi
  40116a:	31 c0                	xor    %eax,%eax
  40116c:	e8 bf fe ff ff       	callq  401030 <printf@plt>
  401171:	8b 74 24 04          	mov    0x4(%rsp),%esi
  401175:	83 e6 03             	and    $0x3,%esi
  401178:	bf 04 20 40 00       	mov    $0x402004,%edi
  40117d:	31 c0                	xor    %eax,%eax
  40117f:	e8 ac fe ff ff       	callq  401030 <printf@plt>
```

Reply:
Oh, for my clang 10.0, 
``` asm
## %bb.0:
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        xorl    %edi, %edi
        callq   _f
        xorl    %edi, %edi
        callq   _f
        xorl    %eax, %eax
        popq    %rbp
        retq

```

```
void f(int v) {}
int main() {
    int a;
    int v = a % 4;
    f(v);
    int v2 = a & 3;
    f(v2);
}
```

Reply:
@H-ZeX In my clang 10.0, for the code:

```
int mod1(int a) {
    return a % 4;
}

int mod2(int a) {
    return a & 3;
}

unsigned mod3(unsigned a) {
    return a % 4;
}

unsigned mod4(unsigned a) {
    return a & 3;
}
```

using `gcc -Og -S mod_or_xor.c`:

``` asm
_mod1:                                  ## @mod1
	.cfi_startproc
## %bb.0:
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset %rbp, -16
	movq	%rsp, %rbp
	.cfi_def_cfa_register %rbp
	movl	%edi, %eax
	sarl	$31, %eax
	shrl	$30, %eax
	addl	%edi, %eax
	andl	$-4, %eax
	subl	%eax, %edi
	movl	%edi, %eax
	popq	%rbp
	retq
	.cfi_endproc
                                        ## -- End function
	.globl	_mod2                   ## -- Begin function mod2
	.p2align	4, 0x90
_mod2:                                  ## @mod2
	.cfi_startproc
## %bb.0:
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset %rbp, -16
	movq	%rsp, %rbp
	.cfi_def_cfa_register %rbp
	andl	$3, %edi
	movl	%edi, %eax
	popq	%rbp
	retq
	.cfi_endproc
                                        ## -- End function
	.globl	_mod3                   ## -- Begin function mod3
	.p2align	4, 0x90
_mod3:                                  ## @mod3
	.cfi_startproc
## %bb.0:
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset %rbp, -16
	movq	%rsp, %rbp
	.cfi_def_cfa_register %rbp
	andl	$3, %edi
	movl	%edi, %eax
	popq	%rbp
	retq
	.cfi_endproc
                                        ## -- End function
	.globl	_mod4                   ## -- Begin function mod4
	.p2align	4, 0x90
_mod4:                                  ## @mod4
	.cfi_startproc
## %bb.0:
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset %rbp, -16
	movq	%rsp, %rbp
	.cfi_def_cfa_register %rbp
	andl	$3, %edi
	movl	%edi, %eax
	popq	%rbp
	retq
	.cfi_endproc
```

If we're using uint, I think they're the same.

Reply:
With all due respect, I think they are the same. It is just a code-style problem. We should forget about it. XD